tracing-core/src/callsite.rs

```diff
@@ -169,3 +187,159 @@ impl Hash for Identifier {
         (self.0 as *const dyn Callsite).hash(state)
     }
 }
+
+// ===== impl Registration =====
+
+impl<T> Registration<T> {
+    /// Construct a new `Registration` from some `&'static dyn Callsite`
+    pub const fn new(callsite: T) -> Self {
+        Self {
+            callsite,
+            next: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+}
+
+impl fmt::Debug for Registration {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Registration").finish()
+    }
+}
+
+// ===== impl LinkedList =====
+
+/// An intrusive atomic push only linked-list.
+struct LinkedList {
+    head: AtomicPtr<Registration>,
+}
+
+impl LinkedList {
+    fn new() -> Self {
+        LinkedList {
+            head: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+
+    fn for_each(&self, mut f: impl FnMut(&'static dyn Callsite)) {
+        let mut head = self.head.load(Ordering::Acquire);
+
+        while !head.is_null() {
+            let reg = unsafe { &*head };
+            f(reg.callsite);
+
+            head = reg.next.load(Ordering::Acquire);
+        }
+    }
+
+    fn push(&self, cs: &'static dyn Callsite) {
+        let mut head = self.head.load(Ordering::Acquire);
+
+        loop {
+            let registration = cs.registration() as *const _ as *mut _;
+
+            cs.registration().next.store(head, Ordering::Release);
+
+            match self.head.compare_exchange(
+                head,
+                registration,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            ) {
+                Ok(old) => {
+                    assert_ne!(
+                        cs.registration() as *const _,
+                        old,
+                        "Attempting to push a `Callsite` that already exists. \
+                        This will cause an infinite loop when attempting to read from the \
+                        callsite cache. This is likely a bug! You should only need to push a \
+                        `Callsite` once."
+                    );
+                    break;
+                }
+                Err(current) => head = current,
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[derive(Eq, PartialEq)]
+    struct Cs1;
+    static CS1: Cs1 = Cs1;
+    static REG1: Registration = Registration::new(&CS1);
+
+    impl Callsite for Cs1 {
+        fn set_interest(&self, _interest: Interest) {}
+        fn metadata(&self) -> &Metadata<'_> {
+            todo!()
```

Comment:
nit/tioli: slight pref to use `unimplemented!` rather than `todo!` in tests --- `todo!` kind of suggests that we intend to do this later, and is useful for checking if there's still code in a branch that needs to be finished before it merges. these'll show up in `rg todo`, which is not really necessary.

i might prefer
```suggestion
            unimplemented!("not needed for this test")
```
or something.